File: bigdime-libraries/bigdime-hdfs-lib/src/main/java/io/bigdime/libs/hdfs/jdbc/HiveJdbcConnectionFactory.java
package io.bigdime.libs.hdfs.jdbc;

import java.io.IOException;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.Map;

import org.apache.commons.dbcp.BasicDataSource;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.security.UserGroupInformation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

@Component
@Scope("prototype")
public class HiveJdbcConnectionFactory {
	private static final Logger logger = LoggerFactory.getLogger(HiveJdbcConnectionFactory.class);

	private static final String SEMI_COLON = ";";
	private static final String QUESTION_MARK = "?";

	public Connection getConnectionWithKerberosAuthentication(final String driverClassName, final String jdbcUrl,
			final String keytabUser, final String keytabPath, final Map<String, String> hiveConfigurations)
			throws IOException, SQLException, ClassNotFoundException {
		Configuration conf = new Configuration();
		conf.set("hadoop.security.authentication", "Kerberos");
		return getConnectionWithKerberosAuthentication(driverClassName, jdbcUrl, conf, keytabUser, keytabPath,
				hiveConfigurations);
	}

	public Connection getConnectionWithKerberosAuthentication(final String driverClassName, final String jdbcUrl,
			final Configuration conf, final String keytabUser, final String keytabPath,
			final Map<String, String> hiveConfigurations) throws IOException, SQLException, ClassNotFoundException {
		return getConnection0(driverClassName, jdbcUrl, conf, keytabUser, keytabPath, hiveConfigurations);
	}

	private Connection getConnection0(final String driverClassName, final String jdbcUrl, final Configuration conf,
			final String keytabUser, final String keytabPath, final Map<String, String> hiveConfigurations)
			throws IOException, SQLException, ClassNotFoundException {

		String jdbcUrlWithConf = jdbcUrl;

		StringBuilder hiveConfVars = new StringBuilder();
		for (String key : hiveConfigurations.keySet()) {
			hiveConfVars.append(key).append("=").append(hiveConfigurations.get(key)).append(SEMI_COLON);
		}
		if (hiveConfVars.length() > 0) {
			jdbcUrlWithConf = jdbcUrl + QUESTION_MARK + hiveConfVars.toString();
		}

		logger.debug("connecting to db, using kerberos auth",
				"jdbcUrl=\"{}\" driverClassName=\"{}\" keytabUser=\"{}\" keytabPath=\"{}\"", jdbcUrlWithConf,
				driverClassName, keytabUser, keytabPath);

		BasicDataSource datasource = new BasicDataSource();
		datasource.setDriverClassName(driverClassName);
		datasource.setUrl(jdbcUrlWithConf);
		loginUserFromKeytab(conf, keytabUser, keytabPath);

		return datasource.getConnection();

	}

	private void loginUserFromKeytab(final Configuration conf, final String keytabUser, final String keytabPath)
			throws IOException {
		UserGroupInformation.setConfiguration(conf);
		UserGroupInformation.loginUserFromKeytab(keytabUser, keytabPath);

	}

	public Connection getConnection(final String driverClassName, final String jdbcUrl, final String username,
			final String password, final Map<String, String> hiveConfigurations) throws IOException, SQLException {
		return getConnection0(driverClassName, jdbcUrl, username, password, hiveConfigurations);
	}

	private Connection getConnection0(final String driverClassName, final String jdbcUrl, final String username,
			final String password, final Map<String, String> hiveConfigurations) throws IOException, SQLException {

		String jdbcUrlWithConf = jdbcUrl;

		StringBuilder hiveConfVars = new StringBuilder();
		for (String key : hiveConfigurations.keySet()) {
			hiveConfVars.append(key).append("=").append(hiveConfigurations.get(key)).append(SEMI_COLON);
		}
		if (hiveConfVars.length() > 0) {
			jdbcUrlWithConf = jdbcUrl + QUESTION_MARK + hiveConfVars.toString();
		}

		logger.debug("connecting to db, using username and password",
				"jdbcUrl=\"{}\" driverClassName=\"{}\" username=\"{}\" password=\"{}\"", jdbcUrlWithConf,
				driverClassName, username, "*****");

		BasicDataSource datasource = new BasicDataSource();
		datasource.setDriverClassName(driverClassName);
		datasource.setUrl(jdbcUrlWithConf);
		datasource.setUsername(username);
		datasource.setPassword(password);
		return datasource.getConnection();
	}
}

File: bigdime-libraries/bigdime-hdfs-lib/src/main/scala/io/bigdime/libs/hdfs/HiveJdbcConnectionFactory.scala
package io.bigdime.libs.hdfs

import java.io.IOException
import java.sql.{Connection, SQLException}

import org.apache.commons.dbcp.BasicDataSource
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.security.UserGroupInformation
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import scala.collection.JavaConversions._

/**
  * Created by neejain on 2/14/17.
  */

case class HiveConnectionParam(hdfsAuthOption: HDFS_AUTH_OPTION, driverClassName: String, jdbcUrl: String, userName: String, password: String, hiveConfigurations: java.util.Map[String, String])

@Component
@Scope("prototype")
class HiveJdbcConnectionFactory {
  private val logger = LoggerFactory.getLogger(classOf[HiveJdbcConnectionFactory])
  private val SEMI_COLON = ";"
  private val QUESTION_MARK = "?"

  @throws[IOException]
  @throws[SQLException]
  @throws[ClassNotFoundException]
  def getConnection(connectionParam: HiveConnectionParam): Connection = {
    if (connectionParam.hdfsAuthOption eq HDFS_AUTH_OPTION.KERBEROS) {
      val con = getConnectionWithKerberosAuthentication(connectionParam.driverClassName, connectionParam.jdbcUrl, connectionParam.userName, connectionParam.password, connectionParam.hiveConfigurations)
      logger.info("connected to db using kerberos")
      con
    }
    else if (connectionParam.hdfsAuthOption eq HDFS_AUTH_OPTION.PASSWORD) {
      val con = getConnection(connectionParam.driverClassName, connectionParam.jdbcUrl, connectionParam.userName, connectionParam.password, connectionParam.hiveConfigurations)
      logger.info("connected to db using password")
      con
    } else throw new IllegalArgumentException("Unknown HDFS_AUTH_OPTION:" + connectionParam.hdfsAuthOption)
  }

  @throws[IOException]
  @throws[SQLException]
  @throws[ClassNotFoundException]
  protected def getConnectionWithKerberosAuthentication(driverClassName: String, jdbcUrl: String, keytabUser: String, keytabPath: String, hiveConfigurations: java.util.Map[String, String]): Connection = {
    val conf = new Configuration
    conf.set("hadoop.security.authentication", "Kerberos")
    getConnectionWithKerberosAuthentication(driverClassName, jdbcUrl, conf, keytabUser, keytabPath, hiveConfigurations)
  }

  @throws[IOException]
  @throws[SQLException]
  @throws[ClassNotFoundException]
  protected def getConnectionWithKerberosAuthentication(driverClassName: String, jdbcUrl: String, conf: Configuration, keytabUser: String, keytabPath: String, hiveConfigurations: java.util.Map[String, String]): Connection = getConnection0(driverClassName, jdbcUrl, conf, keytabUser, keytabPath, hiveConfigurations)

  @throws[IOException]
  @throws[SQLException]
  @throws[ClassNotFoundException]
  private def getConnection0(driverClassName: String, jdbcUrl: String, conf: Configuration, keytabUser: String, keytabPath: String, hiveConfigurations: java.util.Map[String, String]) = {
    var jdbcUrlWithConf = jdbcUrl
    val hiveConfVars = new StringBuilder

    for (key <- hiveConfigurations.keySet) {
      hiveConfVars.append(key).append("=").append(hiveConfigurations.get(key)).append(SEMI_COLON)
    }
    if (hiveConfVars.nonEmpty) jdbcUrlWithConf = jdbcUrl + QUESTION_MARK + hiveConfVars.toString
    logger.debug("connecting to db, using kerberos auth", "jdbcUrl=\"{}\" driverClassName=\"{}\" keytabUser=\"{}\" keytabPath=\"{}\"", jdbcUrlWithConf, driverClassName, keytabUser, keytabPath)
    val datasource = new BasicDataSource
    datasource.setDriverClassName(driverClassName)
    datasource.setUrl(jdbcUrlWithConf)
    loginUserFromKeytab(conf, keytabUser, keytabPath)
    datasource.getConnection
  }

  @throws[IOException]
  private def loginUserFromKeytab(conf: Configuration, keytabUser: String, keytabPath: String) {
    UserGroupInformation.setConfiguration(conf)
    UserGroupInformation.loginUserFromKeytab(keytabUser, keytabPath)
  }

  @throws[IOException]
  @throws[SQLException]
  protected def getConnection(driverClassName: String, jdbcUrl: String, username: String, password: String, hiveConfigurations: java.util.Map[String, String]): Connection = getConnection0(driverClassName, jdbcUrl, username, password, hiveConfigurations)

  @throws[IOException]
  @throws[SQLException]
  private def getConnection0(driverClassName: String, jdbcUrl: String, username: String, password: String, hiveConfigurations: java.util.Map[String, String]) = {
    var jdbcUrlWithConf = jdbcUrl
    val hiveConfVars = new StringBuilder
    for (key <- hiveConfigurations.keySet) {
      hiveConfVars.append(key).append("=").append(hiveConfigurations.get(key)).append(SEMI_COLON)
    }
    if (hiveConfVars.nonEmpty) jdbcUrlWithConf = jdbcUrl + QUESTION_MARK + hiveConfVars.toString
    logger.debug("connecting to db, using username and password", "jdbcUrl=\"{}\" driverClassName=\"{}\" username=\"{}\" password=\"{}\"", jdbcUrlWithConf, driverClassName, username, "*****")
    val datasource = new BasicDataSource
    datasource.setDriverClassName(driverClassName)
    datasource.setUrl(jdbcUrlWithConf)
    datasource.setUsername(username)
    datasource.setPassword(password)
    datasource.getConnection
  }
}
